### PR TITLE
feat: add requires_token_ids interface for sampling params

### PR DESCRIPTION
**Pull request by llsj14**
## Purpose
- Async scheduling currently blocks the delivery of prompt/output token IDs to sampling parameters, which means that, for example, prompt or output token IDs are set to -1 and only updated asynchronously.
- Certain logits processors (such as [thinking budget](https://github.com/vllm-project/vllm/pull/20859))require access to prompt and output token IDs, and therefore an explicit interface is needed to propagate this information to sampling parameters.
- Pooling parameters already expose such an interface, and this PR applies the same interface to sampling parameters for consistency.
- The changes are interface-only and do not affect execution or runtime behavior.

## Test Plan
- This PR is expected to only add an interface layer without impacting the existing execution flow.
- No behavioral changes are introduced, and existing functionality remains unaffected.

## Prior discussion
https://github.com/vllm-project/vllm/pull/20859#discussion_r2674557146

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 6bab1b4b61e18988ab1798581014ec7ddb745758. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds an interface to request token IDs for logits processing and integrates it into batching.
> 
> - Introduces `SamplingParams.requires_token_ids` (default `False`)
> - Propagates the flag in `gpu_input_batch.py` to set `logits_processing_needs_token_ids` per request
> - Updates `SamplingMetadata` construction to include `prompt_token_ids` and `output_token_ids` when any request requires them (in addition to penalties/bad words/logitsprocs)
> - No other logic changes; mirrors existing pooling params behavior
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6bab1b4b61e18988ab1798581014ec7ddb745758. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds an explicit flag so logits processing can request token IDs under async scheduling, mirroring pooling behavior.
> 
> - Introduces `SamplingParams.requires_token_ids` (default `False`)
> - Propagates per-request flag to `gpu_input_batch` (`logits_processing_needs_token_ids`) during add, swap, and condense
> - Updates `SamplingMetadata` construction to include `prompt_token_ids`/`output_token_ids` when any request sets the flag (in addition to penalties, bad words, or logits processors)
> - No other logic changes
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4c4bc808231791020bf306b2b55642162f226ef1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces an opt-in interface for logits processing to request token IDs and wires it into batching.
> 
> - Adds `SamplingParams.requires_token_ids` (default `False`) to indicate token IDs are needed for logits processing (parity with `PoolingParams`)
> - Propagates per-request flag in `gpu_input_batch.py`: set/read `logits_processing_needs_token_ids` on add, swap, and condense
> - Updates `SamplingMetadata` construction to include `prompt_token_ids` and `output_token_ids` when any request sets the flag (in addition to penalties/bad_words/logitsprocs)
> - Default behavior unchanged unless the flag is set
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3ec65def68bb5bd6a797611d97ef12ec217bf7fa. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: vllm/v1/worker/gpu_input_batch.py [hunk 1]
```diff
@@ -410,6 +410,10 @@ def add_request(
                 self.bad_words_token_ids[req_index] = (
                     sampling_params.bad_words_token_ids
                 )
+
+            self.logits_processing_needs_token_ids[req_index] = (
+                sampling_params.requires_token_ids
+            )
```
[the remark this answers is not in the document]
Comment (llsj14) [reply vllm/v1/worker/gpu_input_batch.py:416]: I added some logic according to this review

**Comment by njhill** [comment]
@llsj14 `output_token_ids` will already be available to any custom logits processors. The logic already takes this into account: https://github.com/vllm-project/vllm/blob/f243abc92d805b366ca00911d6ae5572faa39bff/vllm/v1/worker/gpu_model_runner.py#L507-L509

**Comment by rishitdholakia13** [comment]
Hey @njhill , a quick follow up on the PR ! Would be great if you could provide your views/approvals on the PR. It’s needed for a dependent integration [PR](https://github.com/vllm-project/vllm/pull/20859). This is blocking some of the workloads on Cohere's end and we would like to get the thinking budget PR merged soon, to unblock us. Thank you !

**Comment by njhill** [comment]
@llsj14 sorry again for the delay. Let me try to explain the hesitation:

The current LogitsProcessor framework is not really in a fully working state - in particular it doesn't work properly with preemptions or pipeline parallelism. The way that logit processors that require access to input/output ids  are handled is also clunky and inefficient/inconsistent (including for built-in params like penalties)... in particular w.r.t. how the ids are managed between cpu and gpu and in the context of async scheduling.

We plan to soon re-add support for LogitProcessors with a better interface and with these issues addressed, as part of the V2 Model Runner work, and don't want to invest the effort to refactor the V1 logits processor logic and token handling.

If we are adding thinking token budget as a built-in logits processor in V1, it's probably best to handle for now in the same way as bad_words / penalties are, i.e. hardcoded addition to the `needs_output_token_ids` check in `InputBatch._make_sampling_metadata`, rather than trying to add more generic handling.

I don't actually understand the `requires_token_ids` property added here since it always returns False, and I wasn't aware that we treat the SamplingParams class itself as custom-extensible.
